cloud/cloud_env_impl.cc

```diff
@@ -1840,6 +1843,45 @@ Status CloudEnvImpl::RollNewEpoch(const std::string& local_dbname) {
   return Status::OK();
 }
 
+Status CloudEnvImpl::UploadLocalCloudManifest(const std::string& local_dbname,
+                                              const std::string& cookie) {
+  assert(HasDestBucket());
+
+  std::string current_epoch = cloud_manifest_->GetCurrentEpoch();
+  // upload the manifest file
+  auto st = GetStorageProvider()->PutCloudObject(
+      ManifestFileWithEpoch(local_dbname, current_epoch), GetDestBucketName(),
+      ManifestFileWithEpoch(GetDestObjectPath(), current_epoch));
+  if (!st.ok()) {
+    return st;
+  }
+
+  // upload the cloud manifest file corresponds to cookie (i.e., CLOUDMANIFEST-cookie)
+  return GetStorageProvider()->PutCloudObject(
+      MakeCloudManifestFile(local_dbname, cookie), GetDestBucketName(),
+      MakeCloudManifestFile(GetDestObjectPath(), cookie));
+}
+
+Status CloudEnvImpl::ApplyLocalCloudManifestDelta(const std::string& local_dbname,
+    const std::string& new_cookie,
+    const CloudManifestDelta& delta) {
+  std::string old_epoch = cloud_manifest_->GetCurrentEpoch();
+  const auto& fs = GetBaseEnv()->GetFileSystem();
+  auto io_st = CopyFile(fs.get(), ManifestFileWithEpoch(local_dbname, old_epoch),
+                     ManifestFileWithEpoch(local_dbname, delta.epoch),
+                     0 /* size */, true /* use_fsync */,
+                     nullptr /* io_tracer */, Temperature::kUnknown);
```

Comment:
That's a good point. Thinking more about this.

Reply:
So if we make the local manifest filename just MANIFEST, then we don't need to do anything about local MANIFEST when rolling the epoch/CloudManifest. We just need to update local CloudManifest  with the new epoch, upload MANIFEST to s3 as MANIFEST-new_epoch and upload  CLOUDMANIFEST-cookie. That seems to be able to indeed solve the issue.

Reply:
Exactly! I don't really know why we added -suffix to the local MANIFEST filename, I can't think of a single good reason to do it.

Reply:
1. You can't write to the existing `MANIFEST` file because you might get in the corrupted state if you crash mid-write. You can, however, write to `MANIFEST.tmp` and then do an atomic rename `MANIFEST.tmp -> MANIFEST`. I don't know why RocksDB decided to do this atomic switch through the `CURRENT` file instead of the atomic rename. Possibly to support filesystems that don't have a rename operation? Not a concern for us.
2. Today, yes, we write to the same `MANIFEST-epoch` file with the technique from (1).

Reply:
I see. since we are already writing to the same MANIFEST-epoch file, I guess there shouldn't be any issues with writing to `MANIFEST` file (with renaming of course).

Reply:
BTW, do we have concerns if the local `MANIFEST`/`CLOUDMANIFEST` file is not consistent with the one in the cloud? This can happen if we update local file correctly, but uploading to s3 fails.

To me, looks like this can be an issue. We always read from local storage first: https://github.com/rockset/rocksdb-cloud/blob/master/cloud/cloud_env_impl.cc#L172

Also, reopen db won't fix this issue?

Reply:
nvm. I think this is fine. We have the same issue today.

Reply:
I'll create a separate PR for this and gradually roll it out in prod.